docs/data-sources/aws_transit_gateway_attachment.md

```diff
@@ -0,0 +1,51 @@
+---
+page_title: "hcp_aws_transit_gateway_attachment Data Source - terraform-provider-hcp"
+subcategory: ""
+description: |-
+  The AWS Transit gateway attachment data source provides information about an existing Transit gateway attachment.
+---
+
+# Data Source `hcp_aws_transit_gateway_attachment`
+
+The AWS Transit gateway attachment data source provides information about an existing Transit gateway attachment.
+
+## Example Usage
+
+```terraform
+data "hcp_aws_transit_gateway_attachment" "test" {
+  hvn_id                        = var.hvn_id
+  transit_gateway_attachment_id = var.transit_gateway_attachment_id
+}
+```
+
+## Schema
+
+### Required
+
+- **hvn_id** (String) The ID of the HashiCorp Virtual Network (HVN).
+- **transit_gateway_attachment_id** (String) The ID of the Transit gateway attachment.
+
+### Optional
+
+- **id** (String) The ID of this resource.
+- **state** (String) The state of the Transit gateway attachment.
```

Comment:
The `state` field used here was copied from the internal implementation, but I think it should change to a `wait_for_active_state` flag, or `desired_state`. More detail below:

For this use case, I found some prior discussion from the existing AWS and Google providers:
- https://github.com/hashicorp/terraform-provider-aws/pull/3485
- https://github.com/hashicorp/terraform-provider-google/pull/4797

In short, there seems to be two concerns with using the field `state`:
1. It is merged with the computed `state` of the resource, which can make things behave strangely when trying to resolve whether the `state` has been set by the user or the service ([ref](https://github.com/hashicorp/terraform-provider-google/pull/4797#pullrequestreview-353254440)).
2. It is ambiguous whether the field is meant to wait for the specified state, or filter by the specified state. The [AWS provider is already using state fields as filters](https://github.com/hashicorp/terraform-provider-aws/blob/main/aws/data_source_aws_vpc_peering_connection.go#L120-L130), so it could be confusing if users expect that same usage here.

I came across these two general solutions to the problem that I think we should consider:
1. `wait_for_active_state = "true"` ([ref](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/ecs_service#wait_for_steady_state))
2. `desired_state = "ACTIVE"` ([ref](https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/compute_instance#desired_status))

Reply:
Iiiinteresting conundrum 🤔 I'm leaning towards #1, unless we can think of any other state the user might want to wait for (maybe PENDING ACCEPTANCE). I've got some questions though.

> It is merged with the computed `state` of the resource
Do you mean terraform a different state that TF is tracking?

Say we go with #1. What happens if the resource never gets set to active? How can we find out its state?

Reply:
What are the current different states of a TGW connection? 
Also I assume this is different from the state discussion over in: https://github.com/hashicorp/terraform-provider-hcp/pull/59

Reply:
> Iiiinteresting conundrum 🤔 I'm leaning towards #1, unless we can think of any other state the user might want to wait for (maybe PENDING ACCEPTANCE). I've got some questions though.
> 
> > It is merged with the computed `state` of the resource
> > Do you mean terraform a different state that TF is tracking?
> 
> Say we go with #1. What happens if the resource never gets set to active? How can we find out its state?

Our polling logic should account for any "terminal" state (errored, etc that we can never recover from) and throw an error. 

I prefer the `wait_for_active_state` approach of option 1 as it puts less burden on the user to inspect every possible state. It would also align with doing the same for peering connections as discussed in https://hashicorp.atlassian.net/browse/HCPE-871

Reply:
For reference, possible states for TGW attachment currently are ([ref](https://github.com/hashicorp/cloud-network/blob/master/proto-public/20200907/tgw_attachment.proto#L47-L55)):
```
UNSET
CREATING
PENDING_ACCEPTANCE
ACCEPTED
ACTIVE
FAILED
REJECTED
DELETING
EXPIRED
```

| unless we can think of any other state the user might want to wait for (maybe PENDING ACCEPTANCE)
Yea, it seems like `PENDING_ACCEPTANCE` and `ACTIVE` would be the two states we would be interested in

| Do you mean terraform a different state that TF is tracking?
There's more context in [this link](https://github.com/hashicorp/terraform-provider-google/pull/4797#pullrequestreview-353254440), but I think the concern is that the `state` could be updated from the server during a READ operation, and then you don't really know anymore what state the user is waiting for.

| Say we go with #1. What happens if the resource never gets set to active? How can we find out its state?
Yea, what @alexmunda mentioned sounds like the right solution in that case: store the state of the resource as it gets updated within the operation.

Reply:
Thanks for the comments everyone. This is a good thread. I agree with the teams viewpoint on the 1st item.

Reply:
In case of going with `wait_for_active_state` approach I would recommend deleting `state` input, otherwise you'll have to deal with edge-cases when `wait_for_active_state = true` and `state` is set to something different.

Reply:
Agreed, thanks for the comments! I'll proceed with a `wait_for_active_state` flag and change `state` to be read-only.